<div class="clearfix" style="padding: 10px; padding-left: 0px">
<a href="http://bombora.com"><img src="https://app.box.com/shared/static/e0j9v1xjmubit0inthhgv3llwnoansjp.png" width="200px" class="pull-right" style="display: inline-block; margin: 5px; vertical-align: middle;"></a>
<h1> Bombora Data Science: <br> *Interview Exam* </h1>
</div>

<img width="200px" src="https://app.box.com/shared/static/15slg1mvjd1zldbg3xkj9picjkmhzpa5.png">

---
# Welcome

Welcome! This notebook contains interview exam questions referenced in the *Instructions* section in the `README.md`—please read that first, *before* attempting to answer questions here.

<div class="alert alert-info" role="alert" style="margin: 10px">
<p style="font-weight:bold">ADVICE</p>
<p>*Do not* read these questions, and panic, *before* reading the instructions in `README.md`.</p>
</div>

<div class="alert alert-warning" role="alert" style="margin: 10px">
<p style="font-weight:bold">WARNING</p>

<p>If using <a href="https://try.jupyter.org">try.jupyter.org</a> do not rely on the server for anything you want to last - your server will be <span style="font-weight:bold">deleted after 10 minutes of inactivity</span>. Save often and rember download notebook when you step away (you can always re-upload and start again)!</p>
</div>


## Have fun!

Regardless of outcome, getting to know you is important. Give it your best shot and we'll look forward to following up!

# Exam Questions

## 1. Algo + Data Structures

# <font color=green>SOLVED</font>

### Q 1.1: Fibionacci
![fib image](https://upload.wikimedia.org/wikipedia/commons/thumb/9/93/Fibonacci_spiral_34.svg/200px-Fibonacci_spiral_34.svg.png)

#### Q 1.1.1
Given $n$ where $n \in \mathbb{N}$ (i.e., $n$ is an integer and $n > 0$), write a function `fibonacci(n)` that computes the Fibonacci number $F_n$, where $F_n$ is defined by the recurrence relation:

$$ F_n = F_{n-1} + F_{n-2}$$

with initial conditions of:

$$ F_1 = 1,  F_2 = 1$$

In [1]:
'''
This method is called 'Iterative Approach'. Here we use the loops to compute the fibonacci number. 
'''
def fibonacci(n):
    # Handle the base case
    if n <= 1:
        return n
    num1,num2 = 1,1
    
    # Iterative for loop
    for _ in range(2,n):
        # Update the valus of num1 and num2 
        num1, num2 = num2, num1+num2
    
    return num2

def main():
    # Test Cases:
    test_cases = [1,2,3,4,5,6,7,8,9,10]
    
    # Call the function for all the test cases and verify
    for n in test_cases:
        ans = fibonacci(n)
        print('Fibonacci of {} is {}'.format(n, ans))
        
# Calling the main driver function
main()

Fibonacci of 1 is 1
Fibonacci of 2 is 1
Fibonacci of 3 is 2
Fibonacci of 4 is 3
Fibonacci of 5 is 5
Fibonacci of 6 is 8
Fibonacci of 7 is 13
Fibonacci of 8 is 21
Fibonacci of 9 is 34
Fibonacci of 10 is 55


#### Q 1.1.2
What's the complexity of your implementation?

#### Time complexity: O(n)
According to number of iterations required in the loop, we would need to iterate n-2 times. Thus, as value of n 
increases the computation will increase linearly. So, the time complexity would be O(n).

#### Space Complexity: O(1)
We need two integers to store the previous two numbers. The space occupied would remain constant here. So, the space complexity
of the program would be O(1).


#### Q 1.1.3
Consider an alternative implementation to compute Fibonacci number $F_n$ and write a new function, `fibonacci2(n)`.

In [2]:
'''
This method is called 'Recursive Approach'. Here, the function makes repeatitive calls to iteself. 
        
'''
def fibonacci2(n):
    # Handling the base case
    if n == 0:
        return 0
    if n == 1:
        return 1
    
    # Make recursive call the itself
    return fibonacci(n-1) + fibonacci(n-2)
    
def main():
    # Test Cases:
    test_cases = [1,2,3,4,5,6,7,8,9,10]
    
    # Call function for all test cases and verify
    for n in test_cases:
        ans = fibonacci(n)
        print('Fibonacci of {} is {}'.format(n, ans))

# Calling the main driver function
main()
    

Fibonacci of 1 is 1
Fibonacci of 2 is 1
Fibonacci of 3 is 2
Fibonacci of 4 is 3
Fibonacci of 5 is 5
Fibonacci of 6 is 8
Fibonacci of 7 is 13
Fibonacci of 8 is 21
Fibonacci of 9 is 34
Fibonacci of 10 is 55


#### Q 1.1.4
What's the complexity of your implementation?

#### Time Complexity: O(2^n)
The computation that we need to do increases expotentially. Because, we need to make a recursive call over the last two numbers.
For example, to calculate O(n) we will need to calculate O(n-1) + O(n-2). O(n-2) will again make a call to compute O(n-3) 
and O(n-4) and so on. Thus, the computations will double as we increase the value of n.


#### Space Complexity: O(n)
We will not need to store any values.The computation happens dynamically using the recursive calls. But memory stack is used as a 
function call stack, which stores sequence of functions that need to be performed. So, the space complexity is O(n).
 

#### Q 1.1.5
What are some examples of optimizations that could improve computational performance?


### Optimization Methods:

#### 1. Using Memoization in Recursive Calls:

Time Complexity: O(n)

A memory cache can be used which can store the previous computation of fibonacci sequence. This will help to avoid the redundant
calculations that are done for the same number. This approach uses dynamic programming. It would help to make the function more effecient. The cache would be checked first before computing values again. The new time complexity would be O(n) as all the values would be computed only once.

Here a memory cache can be a dictionary or an array. 

#### 2. Using Matrix :
Time Complexity: O(logn)

A method called 'Matrix Expotentiation' can be used to reduce the computation effiecieny to O(log(n)). It the uses the base matrix
which can be represented as follows:<br>
    m = [[1,0],
         [1,1]]<br>
Using repeating squares over the matrix 'm' (for example n times, m^n), the result matrix formed would give the value of 
fibonnaci number.



### Q 1.2: Linked List
![ll img](https://upload.wikimedia.org/wikipedia/commons/thumb/6/6d/Singly-linked-list.svg/500px-Singly-linked-list.svg.png)

#### Q 1.2.1
Consider a [singly linked list](https://en.wikipedia.org/wiki/Linked_list), $L$. Write a function `is_palindrome(L)` that detects if $L$ is a [palindrome](https://en.wikipedia.org/wiki/Palindrome), by returning a bool, `True` or `False`.


#### Q 1.2.2
What is the complexity of your implementation?

#### Q 1.2.3
Consider an alternative implementation to detect if L is a palindrome and write a new function, `is_palindrome2(L)`.

#### Q 1.2.4
What's the complexity of this implementation?


#### Q 1.2.5 
What are some examples of optimizations that could improve computational performance?


## 2. Prob + Stats

# <font color=green>SOLVED</font>

### Q 2.1: Finding $\pi$ in a random uniform?
![pi pie img](http://core2.staticworld.net/images/article/2016/03/pi-day-intro-100649273-carousel.idge.jpeg)

Given a uniform random generator $[0,1)$ (e.g., use your language's standard libary to generate random value), write a a function `compute_pi` to compute [$\pi$](https://en.wikipedia.org/wiki/Pi).

### The approach used here to find $\pi$ deals with area of circle and area of square. The explanation of the solutions is as follows:

1. Take a circle of radius 1 with center (0,0) and a square enclosing that circle. The random number generator is restricted to the values between 0 and 1. That's all we need here.
<br>
2. The function is made modular so that the user can decide how many random numbers are to be used to approximate the value of $\pi$
<br>
3. Now, we will focus on the positive values from 0 to 1 i.e. on the positive quarter of the circle and square which lies in first quadrant. So, all random points we get would be positive and will be in first quandrant.
<br>
4. A 'for' loop will generate random point. Number for points that are found inside the circle would be calculated. This is done 
by using Pythagoras law. 
<br>
5. Once all points are found, now we will use area formula. 

   a_square = area of square in 1st quandrant = 1 <br>
   a_circle = area of quarter circle present in 1st quandrant = ($\pi$ $*$ radius $*$ radius)/ 4
   
   num of points inside circle/ num of total points in square =  a_circle/ a_square
   
   Now, a_square and radius is 1. 
   
   num of points inside circle / num of total points in square = $\pi$ / 4
   
   Finally,
   $\pi$ = (4 $*$ num of points inside circle) / num of total points in square
   
   
#### The accuracy of the function gradually increases as we increase the number of random points that are being considered.

In [3]:
#import python libraries required for the execution of the program
import random 

def compute_pi(n):
    radius = 1
    # Take a variable to store how many points are found inside the circle     
    points_inside = 0
    # Take a variable to decide the total number of random numbers to generate
    total_points = n
    
    for _ in range(n):
        # random() gives the random floating point number in the range [0.0, 1.0).
        # X and Y cordinates are taken randomly
        cord_x, cord_y = random.random(), random.random()
        
        # Check if they are inside the circle.
        # Check against '1' because of the pythagoras law
        if (cord_x**2 + cord_y**2) <= 1:
            points_inside += 1
        
    # based on ratio of area of circle and area of square
    value_of_pi = (points_inside * 4) / total_points
    
    return value_of_pi

def main():
    test_cases = [10,100,1000,10000,100000]
    for num_of_points in test_cases:
        print('Using {} unifrom random points, the value of PI was found {}.'.format(num_of_points,compute_pi(num_of_points)))
        
main()

Using 10 unifrom random points, the value of PI was found 2.4.
Using 100 unifrom random points, the value of PI was found 3.68.
Using 1000 unifrom random points, the value of PI was found 3.096.
Using 10000 unifrom random points, the value of PI was found 3.122.
Using 100000 unifrom random points, the value of PI was found 3.13544.


# <font color=green>THANK YOU</font>

### Q 2.2: Making a 6-side die roll a 7?

![reno die image](http://thumbs.ebaystatic.com/images/g/IQ8AAOSwvzRXyagD/s-l225.jpg)

Using a single 6-side die, how can you generate a random number between 1 - 7?

### Q 2.3: Is normality uniform?

![normal and uniform distributions](https://qph.ec.quoracdn.net/main-qimg-f6ed71ed1d0059760fb63db384dcbcca-c)

Given draws from a normal distribution with known parameters, how can you simulate draws from a uniform distribution?

### Q 2.4: Should you pay or should you go?

![coin flip](https://lh5.ggpht.com/iwD6MnHeHVAXNBgrO7r4N9MQxxYi6wT9vb0Mqu905zTnNlBciONAA98BqafyjzC06Q=w300)

Let’s say we play a game where I keep flipping a coin until I get heads. If the first time I get heads is on the nth coin, then I pay you $2^{(n-1)}$ US dollars. How much would you pay me to play this game? Explain.

### Q 2.5: Uber vs. Lyft

![uber vs lyft](http://usiaffinity.typepad.com/.a/6a01347fc1cb08970c01bb0876bcbe970d-pi)

You request 2 UberX’s and 3 Lyfts. If the time that each takes to reach you is IID, what is the probability that all the Lyfts arrive first? What is the probability that all the UberX’s arrive first?